### PR TITLE
Grand rework

### LINKED ISSUE #16: Multimonitor support for "pause when fullscreen"
Make it pause LightBulb only on the monitor which hosts the fullscreen application

### LINKED ISSUE #28: Automatic brightness adjustment
I find myself constantly switching the brightness during day and evening.
Could you add an option to automatically change brightness along with screen temperature?

Also, it would be neat if auto-brightness change could come at separate time with yellow-tint. I'd prefer to have my screen dimmed around 6p.m., when the sun goes down, but the orange filter is needed only at 23p.m. - one hour before bedtime.
I don't know if it's worth the hassle, but I'd really enjoy that kind of setup.

### LINKED ISSUE #29: LightBulb stuck?
Like.. This morning I resumed from standby, it was stuck to night temp. 
Tonight then, I just realized it didn't already triggered. 
Only manually disabling/enabling unstucked the thing, but I wouldn't know if tomorrow it mightn't happen again. 

Is there anyway I can take a memory dump or something to add more detail (I still haven't restarted the program)?

### LINKED ISSUE #32: Lightbulb memory explosion
So.. While I was checking the _other computer_ for #29, I right clicked on LB icon... And nothing happened. 
I tried again, again, again and nothing. 
So I opened task manager and I noticed that LB was like 300MB of memory.... Whatever I thought?
Then I clicked again and.. boy, it skyrocketed to 1.2GB
Then it suddenly went down to 100, last but not least like 200 and it crashed. 
All without never ever showing the settings menu. 

```
Application: LightBulb.exe
Frramework version: v4.0.30319
Description: the process was terminated due to an unhandled exception.
Exception Info: System.OutOfMemoryException
   in System.Windows.Media.Composition.DUCE+Channel.SyncFlush()
   in System.Windows.Interop.HwndTarget.UpdateWindowSettings(Boolean, System.Nullable`1<ChannelSet>)
   in System.Windows.Interop.HwndTarget.UpdateWindowPos(IntPtr)
   in System.Windows.Interop.HwndTarget.HandleMessage(MS.Internal.Interop.WindowMessage, IntPtr, IntPtr)
   in System.Windows.Interop.HwndSource.HwndTargetFilterMessage(IntPtr, Int32, IntPtr, IntPtr, Boolean ByRef)
   in MS.Win32.HwndWrapper.WndProc(IntPtr, Int32, IntPtr, IntPtr, Boolean ByRef)
   in MS.Win32.HwndSubclass.DispatcherCallbackOperation(System.Object)
   in System.Windows.Threading.ExceptionWrapper.InternalRealCall(System.Delegate, System.Object, Int32)
   in System.Windows.Threading.ExceptionWrapper.TryCatchWhen(System.Object, System.Delegate, System.Object, Int32, System.Delegate)
   in System.Windows.Threading.Dispatcher.LegacyInvokeImpl(System.Windows.Threading.DispatcherPriority, System.TimeSpan, System.Delegate, System.Object, Int32)
   in MS.Win32.HwndSubclass.SubclassWndProc(IntPtr, Int32, IntPtr, IntPtr)
   in MS.Win32.UnsafeNativeMethods.SetWindowPos(System.Runtime.InteropServices.HandleRef, System.Runtime.InteropServices.HandleRef, Int32, Int32, Int32, Int32, Int32)
   in System.Windows.Interop.HwndSource.Resize(System.Windows.Size)
   in System.Windows.Interop.HwndSource.OnLayoutUpdated(System.Object, System.EventArgs)
   in System.Windows.ContextLayoutManager.fireLayoutUpdateEvent()
   in System.Windows.ContextLayoutManager.UpdateLayout()
   in System.Windows.UIElement.UpdateLayout()
   in System.Windows.Interop.HwndSource.SetLayoutSize()
   in System.Windows.Interop.HwndSource.set_RootVisualInternal(System.Windows.Media.Visual)
   in System.Windows.Interop.HwndSource.set_RootVisual(System.Windows.Media.Visual)
   in System.Windows.Controls.Primitives.Popup.SetRootVisualToPopupRoot()
   in System.Windows.Controls.Primitives.Popup.CreateWindow(Boolean)
   in System.Windows.Controls.Primitives.Popup.OnIsOpenChanged(System.Windows.DependencyObject, System.Windows.DependencyPropertyChangedEventArgs)
   in System.Windows.DependencyObject.OnPropertyChanged(System.Windows.DependencyPropertyChangedEventArgs)
   in System.Windows.FrameworkElement.OnPropertyChanged(System.Windows.DependencyPropertyChangedEventArgs)
   in System.Windows.DependencyObject.NotifyPropertyChange(System.Windows.DependencyPropertyChangedEventArgs)
   in System.Windows.DependencyObject.UpdateEffectiveValue(System.Windows.EntryIndex, System.Windows.DependencyProperty, System.Windows.PropertyMetadata, System.Windows.EffectiveValueEntry, System.Windows.EffectiveValueEntry ByRef, Boolean, Boolean, System.Windows.OperationType)
   in System.Windows.DependencyObject.InvalidateProperty(System.Windows.DependencyProperty, Boolean)
   in System.Windows.Data.BindingExpressionBase.Invalidate(Boolean)
   in System.Windows.Data.BindingExpression.TransferValue(System.Object, Boolean)
   in System.Windows.Data.BindingExpression.ScheduleTransfer(Boolean)
   in MS.Internal.Data.ClrBindingWorker.NewValueAvailable(Boolean, Boolean, Boolean)
   in MS.Internal.Data.PropertyPathWorker.UpdateSourceValueState(Int32, System.ComponentModel.ICollectionView, System.Object, Boolean)
   in MS.Internal.Data.PropertyPathWorker.OnDependencyPropertyChanged(System.Windows.DependencyObject, System.Windows.DependencyProperty, Boolean)
   in MS.Internal.Data.ClrBindingWorker.OnSourceInvalidation(System.Windows.DependencyObject, System.Windows.DependencyProperty, Boolean)
   in System.Windows.Data.BindingExpression.HandlePropertyInvalidation(System.Windows.DependencyObject, System.Windows.DependencyPropertyChangedEventArgs)
   in System.Windows.Data.BindingExpressionBase.OnPropertyInvalidation(System.Windows.DependencyObject, System.Windows.DependencyPropertyChangedEventArgs)
   in System.Windows.Data.BindingExpression.OnPropertyInvalidation(System.Windows.DependencyObject, System.Windows.DependencyPropertyChangedEventArgs)
   in System.Windows.DependentList.InvalidateDependents(System.Windows.DependencyObject, System.Windows.DependencyPropertyChangedEventArgs)
   in System.Windows.DependencyObject.NotifyPropertyChange(System.Windows.DependencyPropertyChangedEventArgs)
   in System.Windows.DependencyObject.UpdateEffectiveValue(System.Windows.EntryIndex, System.Windows.DependencyProperty, System.Windows.PropertyMetadata, System.Windows.EffectiveValueEntry, System.Windows.EffectiveValueEntry ByRef, Boolean, Boolean, System.Windows.OperationType)
   in System.Windows.DependencyObject.SetValueCommon(System.Windows.DependencyProperty, System.Object, System.Windows.PropertyMetadata, Boolean, Boolean, System.Windows.OperationType, Boolean)
   in System.Windows.DependencyObject.SetValue(System.Windows.DependencyProperty, System.Object)
   in Hardcodet.Wpf.TaskbarNotification.TaskbarIcon.OnToolTipChange(Boolean)
   in Hardcodet.Wpf.TaskbarNotification.Interop.WindowMessageSink.ProcessWindowMessage(UInt32, IntPtr, IntPtr)
   in Hardcodet.Wpf.TaskbarNotification.Interop.WindowMessageSink.OnWindowMessageReceived(IntPtr, UInt32, IntPtr, IntPtr)
   in MS.Win32.UnsafeNativeMethods.IntGetMessageW(System.Windows.Interop.MSG ByRef, System.Runtime.InteropServices.HandleRef, Int32, Int32)
   in MS.Win32.UnsafeNativeMethods.GetMessageW(System.Windows.Interop.MSG ByRef, System.Runtime.InteropServices.HandleRef, Int32, Int32)
   in System.Windows.Threading.Dispatcher.GetMessage(System.Windows.Interop.MSG ByRef, IntPtr, Int32, Int32)
   in System.Windows.Threading.Dispatcher.PushFrameImpl(System.Windows.Threading.DispatcherFrame)
   in System.Windows.Threading.Dispatcher.PushFrame(System.Windows.Threading.DispatcherFrame)
   in System.Windows.Application.RunDispatcher(System.Object)
   in System.Windows.Application.RunInternal(System.Windows.Window)
   in System.Windows.Application.Run(System.Windows.Window)
   in LightBulb.App.Main()
```

[Dump](https://github.com/Tyrrrz/LightBulb/files/1067580/crash.7z.pdf)

### LINKED ISSUE #39: Process whitelist

### LINKED ISSUE #42: Enable/disable hotkey does not work when smoothing is off
Thanks for the great program, Fl.ux's hotkey disable only works for 1 hour at time.

I like to use Lightbulb when am programming at all times of the day. But I can switch to design very quickly. I have it set all the way down. As such the transition takes a whole 6 seconds when I toggle :) 

Just being nit picky, but I would happy with more imitate transition.

### LINKED ISSUE #55: Extend colour temperature range
AMD drivers go up to 10k for instance, my screen is very warm by default so I use 7000k to make it the equivalent of 6500k

Lightbulb limits this to 6600k so I have a yellowish tint during the day

### LINKED ISSUE #59: Add manual lat/lon selectors in GUI
For computers that are on VPN networks or otherwise private networks, the automatically detected latitude/longitude may not be accurate. It would be useful in this case to allow users to set this value manually.

### LINKED ISSUE #66: Disable to 'daytime' setting
Would be nice to have an option so that when you choose to disable lightbulb, it reverts to your daytime/brightest setting and doesn't disable it entirely.

### LINKED ISSUE #73: No Option to set to auto start on login
It would be nice to have an option on the context menu so that LightBulb would start automatically when user logs in to windows.

### LINKED ISSUE #90: Allow gamma control down to 1000K?
For circadian rhythm disorders, removing all blue and even yellow light might be necessary to avoid melatonin suppression.

Currently, the app (even by configuring the config file manually in AppData) can only go down to 2100K, which is still yellowish with some blue lights.

Would it be possible to allow the temperature to be turned down to 1000K or even lower? The goal being to have a mostly reddish screen.

### LINKED ISSUE #95: Switches to Daylight Saving Time at 2 am UTC instead of local
Windows 7 clock advances to 6:00 PM local time and my monitor pops back to 6500 °K. Lightbulb is now telling me the sun sets one hour later than it does. We are still on Standard Time until 2:00 AM tomorrow.

### LINKED ISSUE #98: Upgrade to .NET Core 3
.NET Core 3 has performance improvements that .NET Framework 4.8 will not be getting. Migration will also be necessary prior to .NET 5.

### LINKED ISSUE #102: Color reverts when display wakes
After I wake my monitors, the color is reset to daylight. For example, when I walk away and Windows puts the displays to sleep. If I use the Preview (play button) in LightBulb, then it returns to normal operation.

Windows 10 on a Dell laptop, docked in the Dell E-Port Replicator with 2 Dell monitors driven by DVI connection. Laptop monitor not in use.

Maybe related to #53 and #86. Let me know if I can assist with troubleshooting.

Thanks for the app, BTW!  I switched from f.lux for better control over the schedule and more linear color change. Other than this bug, it is perfect for me :)

### PR DESCRIPTION
**Pull request by Tyrrrz**
LightBulb is a rather popular blue light filter but it hasn't been significantly updated in a long time.

Current version of LightBulb was designed with simplicity in mind but this design hinders adding new features that the users want. The idea is to rework LightBulb, both the back- and frontend, to make it more scalable in regards to future improvements. This is a step away from the very simplistic UI that LightBulb used to have, however the aesthetic hopefully will remain the same.

This PR will address these issues:
Closes #28 
Closes #55 
Closes #59 
Closes #66 
Closes #73 
Closes #95 
Closes #98 
Closes #90 

Due to a large amount of changes, it's likely (although not guaranteed, will need to be tested) that it will address these issues:
Maybe affects #102
Maybe affects #42 
Maybe affects #32
Maybe affects #29 
Maybe affects #99

### CODE REVIEW

**Comment by Tyrrrz** [comment]
.NET Core 3 support depends on:
https://github.com/hardcodet/wpf-notifyicon/issues/4
https://github.com/MaterialDesignInXAML/MaterialDesignInXamlToolkit/milestone/25

**Comment by mirh** [comment]
I'm getting this error https://github.com/dotnet/core-setup/issues/6189

**Comment by Tyrrrz** [comment]
This requires you to install .NET Core 3 runtime from here https://dotnet.microsoft.com/download/thank-you/dotnet-runtime-3.0.0-windows-x64-desktop-runtime-installer

Does it work if you do it?

**Comment by mirh** [comment]
Installing .net core is another problem I had, and that it wouldn't hurt to be better handled in the program and in the documentation. 

But I'm getting that error after having installed it.

**Comment by Tyrrrz** [comment]
This is still not meant for public consumption yet. I'll take a look at the issue you mentioned.

**Comment by Tyrrrz** [comment]
@mirh the issue you mentioned should be fixed, you can try again, although it's still not production ready

**Comment by mirh** [comment]
*It woooorks* \o/
But it's telling me that I don't have an unlocked gamma range. 

I had already set `HKLM\SOFTWARE\Microsoft\Windows NT\CurrentVersion\ICM\\GdiICMGammaRange` long time ago, but somehow you are only checking for HKCU (which.. are you even sure is an actually working path?)

**Comment by Tyrrrz** [comment]
I'll take a look

**Comment by Tyrrrz** [comment]
You are right, fixed

**Comment by mirh** [comment]
Hot damn, it's just so comfy and fine. 
The UI is just so top notch (even though perhaps, you could put the K of the temperature outside of the input box?)

But you need administrator privileges to write to HKLM

**Comment by Tyrrrz** [comment]
Yeah I've also noticed the issue with the registry, will have to probably perform this out of process, no problem.

**Comment by mirh** [comment]
I don't know [how](https://stackoverflow.com/questions/3925065/correct-way-to-deal-with-uac-in-c-sharp) [much](https://www.victorhurdugaci.com/using-uac-with-c-part-3) .net core has kept, but it [should](https://stackoverflow.com/questions/42159066/dotnet-core-app-run-as-administrator#comment96451356_47398733) be possible to ask in-process for admin permissions...

Otherwise a .reg file or a batch like this launched in the normal shell should automatically do it for the user

    REG ADD "HKLM\Software\Microsoft\Windows NT\CurrentVersion\ICM" /v GdiICMGammaRange /t REG_DWORD /d 256

---
Anyway
```
Application: LightBulb.exe
CoreCLR Version: 4.700.19.46205
.NET Core Version: 3.0.0
Description: The process was terminated due to an unhandled exception.
Exception Info: System.Net.Http.HttpRequestException: The requested name is valid, but no data of the requested type was found.
 ---> System.Net.Sockets.SocketException (11004): The requested name is valid, but no data of the requested type was found.
   at System.Net.Http.ConnectHelper.ConnectAsync(String host, Int32 port, CancellationToken cancellationToken)
   --- End of inner exception stack trace ---
   at System.Net.Http.ConnectHelper.ConnectAsync(String host, Int32 port, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.ConnectAsync(HttpRequestMessage request, Boolean allowHttp2, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.CreateHttp11ConnectionAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.GetHttpConnectionAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken)
   at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.DecompressionHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.FinishSendAsyncUnbuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)
   at Onova.Services.GithubPackageResolver.GetPackageVersionUrlMapAsync()
   at Onova.Services.GithubPackageResolver.GetPackageVersionsAsync()
   at Onova.UpdateManager.CheckForUpdatesAsync()
   at LightBulb.Services.UpdateService.CheckPrepareUpdateAsync() in C:\projects\lightbulb\LightBulb\Services\UpdateService.cs:line 23
   at LightBulb.ViewModels.RootViewModel.<.ctor>b__35_5() in C:\projects\lightbulb\LightBulb\ViewModels\RootViewModel.cs:line 124
   at System.Threading.Tasks.Task.<>c.<ThrowAsync>b__139_1(Object state)
   at System.Threading.QueueUserWorkItemCallback.<>c.<.cctor>b__6_0(QueueUserWorkItemCallback quwi)
   at System.Threading.ExecutionContext.RunForThreadPoolUnsafe[TState](ExecutionContext executionContext, Action`1 callback, TState& state)
   at System.Threading.QueueUserWorkItemCallback.Execute()
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
```

**Comment by Tyrrrz** [comment]
@mirh yeah I'll get to it later, currently working on some other things. Just click "close" on that dialog for now, when you are testing.

**Comment by mirh** [comment]
Ehrm, that's not a dialog. It's the crash report.

**Comment by Tyrrrz** [comment]
#16 and #39 were descoped (will do them outside of this PR).

**Comment by Tyrrrz** [comment]
@mirh I was talking about extended gamma range dialog

**Comment by mirh** [comment]
MFW I wake the computer and nothing crashed. 
Also good jesus, the brightness and sundial UI. Crazy, stunning, awesome.

**Comment by Tyrrrz** [comment]
Hey, thanks a lot!

I've merged this but I will keep doing more updates to master in the upcoming days, now that this is pretty stable. If you have time, please check CI builds, I really appreciate your continuous quality assurance. :)

Also, if you need to reach me, my Discord is Tyrrrz#5447

**Comment by aure2006** [comment]
CI build 6e28ef75 won't start for me. Tried the portable and installed version. Windows 7 x64.

**Comment by Tyrrrz** [comment]
Did you download .NET Core 3 runtime from https://dotnet.microsoft.com/download/thank-you/dotnet-runtime-3.0.0-windows-x64-desktop-runtime-installer ?

**Comment by NomDeMorte** [comment]
TL;DR at the bottom :)

I wish I'd noticed the CI link in the readme before, I've been checking this page religiously for a month, looking forward to a new release. Grabbed the latest build as of today, hooray! (Pro tip: Don't do this at 0330 MY EYES ARE BURNING lol)

I downloaded the runtime linked above, and the app failed to start. 

Logs show the following:

> Description: A .NET Core application failed.
> Application: LightBulb.exe
> Path: C:\Program Files\LightBulb\LightBulb.exe
> Message: A fatal error occurred. The required library hostfxr.dll could not be found.
> If this is a self-contained application, that library should exist in [C:\Program Files\LightBulb\].
> If this is a framework-dependent application, install the runtime in the global location [C:\Program Files\dotnet] or use the DOTNET_ROOT environment variable to specify the runtime location or register the runtime location in [HKLM\SOFTWARE\dotnet\Setup\InstalledVersions\x64\InstallLocation].

Indeed, the hostfxr.dll file is not installed in either the lightbulb or dotnet directories.

According to this issue: https://github.com/dotnet/cli/issues/3907 this issue is worked around by having installed VS, so I would guess that's why it is not effecting you.

Following the pre-requisites mentioned in that issue, I found myself on the following page: https://dotnet.microsoft.com/download/dotnet-core/3.0/runtime

This lists **two** requisite downloads, the second of which is the one linked above. The first download, installed the missing DLL, and Lightbulb started.


TL;DR

Not sure that we need this: https://dotnet.microsoft.com/download/thank-you/dotnet-runtime-3.0.0-windows-x64-desktop-runtime-installer

We DEFINITELY need this: https://dotnet.microsoft.com/download/thank-you/dotnet-runtime-3.0.0-windows-x64-installer

It works now. Oooh, pretty!

**Comment by Tyrrrz** [comment]
@NomDeMorte thank you for letting me know. I was under the impression that desktop runtime includes the base runtime, but I guess not. I will make sure to include these instructions when v2 is released. :)

**Comment by NomDeMorte** [comment]
> @NomDeMorte thank you for letting me know. I was under the impression that desktop runtime includes the base runtime, but I guess not. I will make sure to include these instructions when v2 is released. :)

You're very welcome and thanks for the great work. Already enjoying my 66% brightness at night, and my altered day/night times :)

I do wonder, based on the event log, if perhaps you might just be able to package the DLL with the application:

> Message: A fatal error occurred. The required library hostfxr.dll could not be found.
> If this is a self-contained application, that library should exist in [C:\Program Files\LightBulb].

If you would like, I could try to un-install the dotnet stuff I just installed, and copy that DLL to the lightbulb directory to see if it works? (I assume it is not so easy for you to test, given that you have the full dev environment in place there).... But perhaps I am mistaken. Let me know if I can help in any way.

**Comment by Tyrrrz** [comment]
I think it's just the first dependency it failed to find and there's more. There's a way to make the app self-contained so you don't have to install the runtime but I'm not a fan of that because it makes the size much higher, i.e. ~200mb.

If we make sure the users know what they need to download, I think it should be fine.

**Comment by NomDeMorte** [comment]
Sounds fine.

Slightly offtopic but I wanted to note, that the rework is using less resources on both GPU and CPU. Not sure if that was intentional, but there it is. Really enjoying this!

A couple of items of minor consequence (perhaps these should be in a separate issue?)

When the app is running, the window position of the settings dialog is remembered, but if I close Lightbulb and re-open it, the settings dialog goes back to the centre of the screen. Would be nice if it remembered where I left it. Not really a big deal, since I practically never change it.

Speaking of re-positioning the settings dialog, the tooltip which pops up below the buttons, can overlay the buttons, if the dialog is at the bottom of the screen:
image

HTH!

**Comment by Tyrrrz** [comment]
>Slightly offtopic but I wanted to note, that the rework is using less resources on both GPU and CPU. Not sure if that was intentional, but there it is. Really enjoying this!

It's probably a mixture of my optimizations and .NET Core. Good news eitherway.

>When the app is running, the window position of the settings dialog is remembered, but if I close Lightbulb and re-open it, the settings dialog goes back to the centre of the screen. Would be nice if it remembered where I left it. Not really a big deal, since I practically never change it.

It didn't use to do it before either :) It's probably better to create a separate issue for that

**Comment by Tyrrrz** [comment]
.NET Core runtime should *hopefully* be installed by the latest installer.

**Comment by mirh** [comment]
It would be even nicer if the executable alone could warn you gracefully it is missing, rather than just crashing badly

**Comment by Tyrrrz** [comment]
Unfortunately that's not possible

**Comment by aure2006** [comment]
2.0 still isn't working for me (won't start and doesn't show any errors). No other lightbulb instance was running when attempting to run lightbulb 2.0.

What I tried:
Reinstalling Windows 7 x64
LightBulb.zip
LightBulb-Installer.exe
Latest graphics drivers
.NET Core 3.1.0 desktop runtime (both x86 and x64)
.NET Core 3.0.1 desktop runtime (both x86 and x64)
.NET Core 3.0.1 runtime (both x86 and x64)
Restarting computer after installing all the .NET Core stuff

It's probably a missing windows update.

**Comment by Tyrrrz** [comment]
@aure2006 can you check the Windows event log? Press Start and type "event viewer" and then go to Windows logs->Application. See if there might be something related.

**Comment by aure2006** [comment]
Found the fix and added it to my previous comment.

**Comment by Tyrrrz** [comment]
Great!
